Make C10_NODISCARD macro more portable for nvcc+clang.

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#20324 Make C10_NODISCARD macro more portable for nvcc+clang.**

Fixes #13118.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D15359317](https://our.internmc.facebook.com/intern/diff/D15359317)

ezyang committed:
Make C10_NODISCARD macro more portable for nvcc+clang.

Fixes #13118.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

mrshenli commented on c10/macros/Macros.h:
```diff
@@ -36,18 +36,32 @@
-#define C10_NODISCARD [[clang::warn_unused_result]]
-#endif
+# if __has_cpp_attribute(clang::warn_unused_result)
+#  ifndef __CUDACC__
```
Is this `if` necessary? Isn't it always true that `__CUDACC__ ` is not defined when entering this branch? Or maybe I don't understand what `defined(__CUDACC__)` does.

ezyang replied:
No, it's not necessary. I'll remove it.

ezyang committed:
Update on "Make C10_NODISCARD macro more portable for nvcc+clang."

Make C10_NODISCARD macro more portable for nvcc+clang.

Fixes #13118.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

gh-metadata: pytorch pytorch 20324 gh/ezyang/125/head

ezyang commented:
Legit Windows failure:

```
21:07:41          C:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build\c10/util/UniqueVoidPtr.h(69): error C2429: attribute 'nodiscard' requires compiler flag '/std:c++latest' [C:\Jenkins\workspace\caffe2-builds\py2-cuda9.0-cudnn7-windows-build\caffe2\caffe2_gpu.vcxproj]
21:07:41          C:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build\c10/core/Allocator.h(96): error C2429: attribute 'nodiscard' requires compiler flag '/std:c++latest' [C:\Jenkins\workspace\caffe2-builds\py2-cuda9.0-cudnn7-windows-build\caffe2\caffe2_gpu.vcxproj]
21:07:41          C:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build\c10/core/TensorOptions.h(171): error C2429: attribute 'nodiscard' requires compiler flag '/std:c++latest' [C:\Jenkins\workspace\caffe2-builds\py2-cuda9.0-cudnn7-windows-build\caffe2\caffe2_gpu.vcxproj]
21:07:41          C:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build\c10/core/TensorOptions.h(181): error C2429: attribute 'nodiscard' requires compiler flag '/std:c++latest' [C:\Jenkins\workspace\caffe2-builds\py2-cuda9.0-cudnn7-windows-build\caffe2\caffe2_gpu.vcxproj]
21:07:41          C:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build\c10/core/TensorOptions.h(190): error C2429: attribute 'nodiscard' requires compiler flag '/std:c++latest' [C:\Jenkins\workspace\caffe2-builds\py2-cuda9.0-cudnn7-windows-build\caffe2\caffe2_gpu.vcxproj]
```

ezyang committed:
Update on "Make C10_NODISCARD macro more portable for nvcc+clang."

Make C10_NODISCARD macro more portable for nvcc+clang.

Fixes #13118.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

gh-metadata: pytorch pytorch 20324 gh/ezyang/125/head

ezyang committed:
Update on "Make C10_NODISCARD macro more portable for nvcc+clang."

Make C10_NODISCARD macro more portable for nvcc+clang.

Fixes #13118.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

gh-metadata: pytorch pytorch 20324 gh/ezyang/125/head

ezyang committed:
Update on "Make C10_NODISCARD macro more portable for nvcc+clang."

Make C10_NODISCARD macro more portable for nvcc+clang.

Fixes #13118.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

gh-metadata: pytorch pytorch 20324 gh/ezyang/125/head
